[RDY] Adds nvim_get_hl_by_name/by_id

teto opened the pull request:
...in order to retrieve highlights.
UIs are also sent new/modified highlights.

I renamed EMPTY_ATTR to HLATTRS_INIT because it's more consistent with the rest of the code and it helps disambiguate the highlight vs attribute vs hlattr confusion. I can revert it though.

teto committed:
Adds nvim_get_hl_by_name/by_id

...in order to retrieve highlights.

Added test/functional/api/highlight_spec.lua
HL_NORMAL is not really a good name, since it's more like an empty attribute than the normal's one.
If one pays attention, syn_cterm_attr2entry is never called with attr=0 because it's always special cased before.
I suggest in subsequent PRs we remove the ATTR_OFF and just insert an EMPTY ATTR/RESET_ATTR/UNINITIALIZED for id 0.

teto commented on src/nvim/ui.c:
```diff
@@ -166,6 +166,88 @@ void ui_event(char *name, Array args)
+  attrs.reverse = mask & (HL_INVERSE | HL_STANDOUT);
+
+  if (use_rgb) {
+    if (aep->rgb_fg_color != normal_fg) {
```
This kind of calls make the export fail for the "Normal" highlight. I'll try to replace "normal_fg" by -1 and see if htis has any adverse impact.

bfredl commented on src/nvim/api/ui_events.in.h:
```diff
@@ -58,7 +58,8 @@ void set_title(String title)
   FUNC_API_SINCE(3);
-
+void highlight_info_set(Array highlights_changed)
+  FUNC_API_SINCE(3);
```
Do you plan to use it in the TUI? Otherwise it should be REMOTE_ONLY and remove unnecessary boilerplate

teto replied:
I plan to use it for TUI, to update the cursor color when the hl changes.

bfredl replied:
Why not include the translated colors in the mode_info dict? We should avoid unnecessary indirections.

teto replied:
I agree but let's keep that for another PR. Meanwhile I will extract the "highlight changed" into separate commit so that you can choose to include it or not (right now all UIs get notifications even when not interested).

bfredl replied:
I'm not against the "highlight changed" event, it will be very useful for other purposes. I'm just saying whenever we implement cursor colors in TUI (in this or next PR doesn't matter) we should do it the right way directly, so there will never be a need for TUI to receive this event.

teto commented:
Travis fails because of:
```
FAILED neovim/neovim/test/functional/autocmd/termclose_spec.lua[0m @ [36m47[0m: [1mTermClose event kills pty job trapping SIGHUP and SIGTERM[0m
...neovim/neovim/test/functional/autocmd/termclose_spec.lua:60: Expected objects to be the same.
Passed in:
(number) 6
Expected:
(number) 4
```
I removed the highlight change notification code to trim down the PR and the most controversial part.

(n)vim used to conflate the 'Normal' hl with the uninitialized attribute (attribute 0).
I've removed this confusion so that an attribute can be generated for the 'Normal' hl too.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -55,6 +56,64 @@ void nvim_command(String command, Error *err)
+  return nvim_get_attr(attrcode, err);
+}
+
+Dictionary nvim_get_attr(Integer attr_id, Error *err)
```
Needs a doc. Would nvim_get_text_attr_by_id be a better name? Or nvim_get_screen_cell_attr_by_id ?

justinmk replied:
Also there's no test for this.

teto replied:
I am not sure how to test that one or if it should be part of the API as there is no public function AFAIK  to map a hl to its attribute (though UIs can receive attribute ids). Attributes are more of a vim internal implementation detail.
What should be done is a way for the UI to combine several hl but I prefer to keep this PR short.

bfredl replied:
I would argue any situation that needs attr combination it should be done nvim side not ui side, with possible exception of ui-managed windows/floats having separate default colors, which can be seen as extension of existing default color management and is much simpler than arbitrary combination

teto commented on src/nvim/ui.c:
```diff
@@ -166,6 +166,88 @@ void ui_event(char *name, Array args)
 }
 
+/// @param[in] aep data to convert
+/// @param[out] out structure that will be sent to UIs
```
gotta fix the doc here :-1:

teto commented on src/nvim/ui.c:
```diff
@@ -166,6 +166,88 @@ void ui_event(char *name, Array args)
+    PUT(hl, "reverse", BOOLEAN_OBJ(true));
+  }
+
+  // todo care, we send none; change order between none/invalid
```
remove

teto commented:
So I moved nvim_get_attr_by_id to nvim/ui.c as it's not of much use and can be confusing if made public. It's tested through its callers nvim_get_hl_by_id / nvim_get_hl_by_name

bfredl commented:
Then `nvim_` prefix should be removed, it is for public functions

teto commented:
done

justinmk commented on test/functional/api/highlight_spec.lua:
```diff
@@ -0,0 +1,54 @@
+describe('highlight api', function()
+
+  before_each(function()
+    clear('--cmd', 'set termguicolors')
```
does it require RGB? why?

teto replied:
I tested the most useful configuration there. I can add without termguicolors as well.

justinmk replied:
Won't it be useful for non-rgb?

teto replied:
The only non-RGB UIs I can think of are terminal based and even those tend to disappear as always more terminals support RGB. I hope that neovim-based UIs won't limit themselves to 16 colors. Anyway that should work even with non-RGB. I can add the tests.

justinmk commented on src/nvim/ui.c:
```diff
@@ -166,6 +166,115 @@ void ui_event(char *name, Array args)
+/// Retrieves attribute description from its id
+///
+/// @param attr_id attribute id
+Dictionary get_attr_by_id(Integer attr_id, Error *err)
```
these new functions are in the UI module but don't have a `ui_` prefix. If they really belong there then they should have that prefix. Not sure if they do belong there though, `hl_` prefix might be better.

teto replied:
should I move it to syntax.c ? or I can merge it with nvim_get_hl_by_id but I believe the current logic separation is clearer.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -55,6 +56,42 @@ void nvim_command(String command, Error *err)
+
+/// Retrieves highlight description from its id
+///
+/// @param hl_id highlight id as returned by hlID()
```
putting bars around tags (`|hlID()|`) will link the tag when the API docs are autogenerated.

justinmk commented on test/functional/api/highlight_spec.lua:
```diff
@@ -0,0 +1,54 @@
+
+  it("nvim_get_hl_by_name", function()
+    local expected_hl = { background = Screen.colors.Yellow,
+                          foreground = Screen.colors.Red }
```
missing tests for "underline", etc.

teto replied:
I considered it was covered by ui tests since they are relying on the same exporting function (to HlAttrs) but I can add that as well.

teto committed:
Increased test coverage for RGB and cterm

teto commented:
I moved the function to syntax.c:hl_get_attr_by_id and added cterm tests.

bfredl commented on test/functional/api/highlight_spec.lua:
```diff
@@ -0,0 +1,60 @@
+    eq({}, nvim("get_hl_by_name", 'Normal'))
+
+    eq(expected_cterm, nvim("get_hl_by_name", 'NewHighlight'))
+    command('set termguicolors')
```
`termguicolors` should only ever affect the TUI. If it affects API calls from an unrelated client (even an external UI?) then that is a bug.

teto replied:
surely we have to be able to convey RGB and non-RGB values through the API, if only to support other TUIs. Do you suggest nvim_get_hl_by_name rely on https://github.com/neovim/neovim/blob/master/src/nvim/api/ui.c#L71 rather than termguicolors or should I change the prototype `nvim_get_hl_by_name (String hl_name, Boolean rgb)` to prevent interference from nvim configuration ?

teto committed:
Changed prototypes to accept a boolean "rgb"

teto commented:
i added an rgb boolean to get a purer function. Is it possible to restart travis plz, looks like a network error.

teto commented:
I believe both failures (qb and travis) are unrelated
```
[ RUN      ] TermClose event triggers when fast-exiting terminal job stops: -- Output to stderr:

Vim: Caught deadly signal 'SIGHUP'

Vim: Finished.

CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:53 (message):

  Running functional tests failed with error: 1.

```

teto commented:
I understand not  to merge code before a release but if https://github.com/neovim/neovim/issues/6428's targeted is still 0.2.1 I can provide subsequent patches once this is merged, 1 of  ~ 6 lines to distinguish between NONE and other colors, the other is also ~10 lines but need some input on how to properly convey hl updates.

bfredl commented on src/nvim/syntax.c:
```diff
@@ -6826,7 +6826,7 @@ int hl_combine_attr(int char_attr, int prim_attr)
     new_en = *char_aep;
   } else {
-    memset(&new_en, 0, sizeof(new_en));
+    new_en = (attrentry_T)ATTRENTRY_INIT;
```
is this not redundant, as `new_en` already has the correct value?

teto replied:
seems redundant indeed

teto committed:
Remove duplicate ATTRENTRY_INIT

justinmk commented on test/functional/api/highlight_spec.lua:
```diff
@@ -0,0 +1,58 @@
+    -- assume there is no hl with id = 30000
+    local err, emsg = pcall(meths.get_hl_by_id, 30000, false)
+    eq(false, err)
+    ok(string.find(emsg, 'Invalid highlight id') ~= nil)
```
I changed this (and the other) to use `eq()` and `string.match()`:

    eq('Invalid highlight id', string.match(emsg, 'Invalid highlight id'))

that gives much nicer failure messages if the test fails.

justinmk commented on src/nvim/syntax.c:
```diff
@@ -6859,6 +6857,7 @@ int hl_combine_attr(int char_attr, int prim_attr)
 
 /// \note this function does not apply exclusively to cterm attr contrary
 /// to what its name implies
+/// \warn don't call it with attr 0 (i.e., the null attribute)
```
@teto was there a reason we can't assert() this?

teto replied:
I think it's safe to assert on this, but it might be even better to remove the +1/-1 all around the code, which IIRC could be done in a backwards compatible way.

justinmk commented:
LGTM. Added some extra tests in 52517321d185

blueyed commented:
https://github.com/neovim/neovim/commit/ba7277cfb4e2556f246446d06b53f3427f28130f seems to break the "new diminactive" I am using:
```vim
    if exists('+winhighlight')
      function! s:configure_winhighlight()
        let ft = &filetype
        let bt = &buftype
        " Check white/blacklist.
        if &diff || (index(['dirvish'], ft) == -1
              \ && (index(['nofile', 'nowrite', 'acwrite', 'quickfix', 'help'], bt) != -1
              \     || index(['startify'], ft) != -1))
          set winhighlight=Normal:MyNormalWin,NormalNC:MyNormalWin
          " echom "normal" winnr() &winhighlight 'ft:'.&ft 'bt:'.&bt
        else
          set winhighlight=Normal:MyNormalWin,NormalNC:MyInactiveWin
          " echom "inactive" winnr() &winhighlight 'ft:'.&ft 'bt:'.&bt
        endif
      endfunction
      augroup inactive_win
        au!
        au ColorScheme * hi link MyInactiveWin ColorColumn | hi link MyNormalWin Normal
        au FileType,BufWinEnter * call s:configure_winhighlight()
        if exists('##OptionSet')
          " TODO: does not work with :Gdiff - diffsplit does not trigger it, too!
          au OptionSet diff call s:configure_winhighlight()
        endif
        " XXX: redraw! should not be necessary, but is?!
        " NOTE: setting cursorline fixes it?!
        " au FocusGained * hi link MyNormalWin Normal | redraw!
        " au FocusLost * hi link MyNormalWin MyInactiveWin | redraw!
        " au FocusGained * hi link MyNormalWin Normal
        " au FocusLost * hi link MyNormalWin MyInactiveWin
        " au FocusGained * echom 'FG: '.&winhighlight
        " au FocusLost * echom 'FL: '.&winhighlight | verb hi MyNormalWin
      augroup END
    endif
```
It now dims the active window, but not the inactive ones after that commit.

blueyed commented:
@teto 
Any idea why this might be reversed now?

teto commented:
in my code, I've tried to make the "Normal" group behave like any other group. I have splitted my PR so that it gets easier to review so I wonder if the problem might be due to this hunk (that should possibly belong to one of my other PR): 
```
// The "Normal" group doesn't need an attribute number
-  if (sgp->sg_name_u != NULL && STRCMP(sgp->sg_name_u, "NORMAL") == 0) {
-    return;
-  }
```
that might badly interact with winhighlight code
```
    if (strncmp("Normal", p, nlen) == 0) {
      w_hl_id_normal = hl_id;
    } else {
      for (hlf = 0; hlf < (int)HLF_COUNT; hlf++) {
        if (strncmp(hlf_names[hlf], p, nlen) == 0) {
          w_hl_ids[hlf] = hl_id;
          break;
        }
      }
      if (hlf == HLF_COUNT) {
        return false;
      }
    }

```
Maybe if you don't special case "Normal" highlight it works
```

      for (hlf = 0; hlf < (int)HLF_COUNT; hlf++) {
        if (strncmp(hlf_names[hlf], p, nlen) == 0) {
          w_hl_ids[hlf] = hl_id;
          break;
        }
      }
      if (hlf == HLF_COUNT) {
        return false;
      }
    }
```

I've got some trouble on my setup so I can't test this. I hope to be able to this weekend (unrelated but my brand new laptop screen refuses to turn on after 2 days of utilization :'( )
